src/services/paras/ParasService.ts

```diff
@@ -0,0 +1,441 @@
+
+		return {
+			at: {
+				hash,
+				height: blockNumber.toString(10),
+			},
+			leasePeriodIndex,
+			endOfLeasePeriod,
+			currentLeaseHolders,
+		};
+	}
+
+	/**
+	 * List all registered paras (parathreads & parachains).
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @returns all the current registered paraIds and their lifecycle status
+	 */
+	async paras(hash: BlockHash): Promise<IParas> {
+		const [{ number }, paraLifecycles] = await Promise.all([
+			this.api.rpc.chain.getHeader(hash),
+			this.api.query.paras.paraLifecycles.entriesAt<ParaLifecycle, [ParaId]>(
+				hash
+			),
+		]);
+
+		const parasPromises = paraLifecycles.map(async ([k, paraLifecycle]) => {
+			const paraId = k.args[0];
+			let onboardingAs: ParaType | undefined;
+			if (paraLifecycle.isOnboarding) {
+				const paraGenesisArgs = await this.api.query.paras.paraGenesisArgs.at<ParaGenesisArgs>(
+					hash,
+					paraId
+				);
+				onboardingAs = paraGenesisArgs.parachain.isTrue
+					? 'parachain'
+					: 'parathread';
+			}
+
+			return {
+				paraId,
+				paraLifecycle,
+				onboardingAs,
+			};
+		});
+
+		return {
+			at: {
+				hash,
+				height: number.unwrap().toString(10),
+			},
+			paras: await Promise.all(parasPromises),
+		};
+	}
+
+	/**
+	 * Calculate the current lease period index.
+	 *
+	 * @param blockHeight current blockheight
+	 * @param leasePeriod duration of lease period
+	 */
+	private currentLeasePeriodIndex(now: BN): BN {
+		const leasePeriod = this.api.consts.slots.leasePeriod as BlockNumber;
+		return now.div(leasePeriod);
+	}
+
+	/**
+	 * The offset into the ending samples of the auction. When we are not in the
+	 * ending phase of the auction we can use 0 as the offset, but we do not return
+	 * that here in order to closely mimic `Auctioneer::is_ending` impl in
+	 * polkadot's `runtime::common::auctions`.
+	 *
+	 * @param now current block number
+	 * @param beginEnd block number of the start of the auction's ending period
+	 */
+	private endingOffset(now: BN, beginEnd: IOption<BN>): IOption<BN> {
+		if (!isSome(beginEnd)) {
+			return null;
+		}
+
+		const afterEarlyEnd = now.sub(beginEnd);
+		if (afterEarlyEnd.lten(0)) {
+			return null;
+		}
+
+		// Once https://github.com/paritytech/polkadot/pull/2848 is merged no longer
+		// need a fallback of 1
+		const sampleLength =
+			(this.api.consts.auctions.sampleLength as BlockNumber) || new BN(1);
+		return afterEarlyEnd.div(sampleLength);
+	}
+
+	/**
+	 * Enumerate in order all the lease sets (SlotRange expressed as a set of
+	 * lease periods) that an `auctions::winning` array covers.
+	 *
+	 * @param leasePeriodIndex
+	 */
+	private enumerateLeaseSets(leasePeriodIndex: BN): number[][] {
+		const leasePeriodIndexNumber = leasePeriodIndex.toNumber();
+		const lPPS =
+			(this.api.consts.auctions.leasePeriodsPerSlot as u32)?.toNumber() ||
+			LEASE_PERIODS_PER_SLOT_FALLBACK;
+
+		const ranges: number[][] = [];
+		for (let start = 0; start < lPPS; start += 1) {
```

Comment:
Out of habit I use `+= 1` because of some weird edge case that some linter I used to use warned about with `++`, but not sure if its relevant. 
Any compelling reason to switch? Convention?

Reply:
Convention yeah, but then again maybe conventions change?